File: Sources/RemoteRift/Lcu/LcuConnection.swift
import Foundation

actor LcuConnection {
  init(
    parser: LockfileParser = LcuLockfileParser(),
    loader: LockfileLoader = LcuLockfileLoader(),
  ) {
    self.parser = parser
    self.loader = loader
  }

  private let parser: LockfileParser
  private let loader: LockfileLoader

  private var lockfileData: LcuLockfileData?

  func getLockfileData() throws -> LcuLockfileData {
    if let lockfileData {
      return lockfileData
    }
    return try refreshLockfileData()
  }

  func refreshLockfileData() throws -> LcuLockfileData {
    let data = try parser.parseLockfile(from: try loader.loadLockfile())
    lockfileData = data
    return data
  }
}

struct LcuLockfileData: Equatable {
  let port: Int
  let password: String
}

enum LcuConnectionError: Error {
  case lockfileMissing
  case lockfileInvalid
}

protocol LockfileParser {
  func parseLockfile(from data: String) throws -> LcuLockfileData
}

class LcuLockfileParser: LockfileParser {
  func parseLockfile(from data: String) throws -> LcuLockfileData {
    let regex = /^(.+):(\d+):(?<port>\d+):(?<password>.*):(.*)$/
    guard let result = data.wholeMatch(of: regex)?.output else {
      throw LcuConnectionError.lockfileInvalid
    }
    return LcuLockfileData(
      port: Int(result.port)!,
      password: String(result.password),
    )
  }
}

protocol LockfileLoader {
  func loadLockfile() throws -> String
}

class LcuLockfileLoader: LockfileLoader {
  func loadLockfile() throws -> String {
    let path = getPlatformPath()
    guard FileManager.default.fileExists(atPath: path) else {
      throw LcuConnectionError.lockfileMissing
    }
    return try String(contentsOfFile: path, encoding: .utf8)
  }

  private func getPlatformPath() -> String {
    #if os(Windows)
      "C:\\Riot Games\\League of Legends\\lockfile"
    #elseif os(macOS)
      "/Applications/League of Legends.app/Contents/LoL/lockfile"
    #else
      fatalError("Unsupported platform — only Windows and macOS are supported.")
    #endif
  }
}

File: packages/core/lib/src/lcu/lcu_connection.dart
import 'dart:io';

class LcuConnection {
  LcuConnection({required this.parser, required this.loader});

  final LcuLockfileParser parser;
  final LcuLockfileLoader loader;

  LcuLockfileData? _lockfileData;

  LcuLockfileData getLockfileData() {
    if (_lockfileData case var data?) {
      return data;
    }
    return refreshLockfileData();
  }

  LcuLockfileData refreshLockfileData() {
    final data = parser.parseLockfile(loader.loadLockfile());
    _lockfileData = data;
    return data;
  }
}

class LcuLockfileParser {
  LcuLockfileData parseLockfile(String data) {
    int? port;
    String? password;

    final regex = RegExp(r'^(.+):(\d+):(\d+):(.*):(.*)$');
    if (regex.firstMatch(data) case var result?) {
      if (result.group(3) case var portValue?) {
        port = int.tryParse(portValue);
      }
      password = result.group(4);
    }

    if (port == null || password == null) {
      throw LcuConnectionError.lockfileInvalid;
    }

    return LcuLockfileData(port: port, password: password);
  }
}

class LcuLockfileLoader {
  String loadLockfile() {
    final path = _getPlatformPath();
    final file = File(path);
    if (!file.existsSync()) {
      throw LcuConnectionError.lockfileMissing;
    }
    return file.readAsStringSync();
  }

  String _getPlatformPath() {
    if (Platform.isWindows) {
      return 'C:\\Riot Games\\League of Legends\\lockfile';
    } else if (Platform.isMacOS) {
      return '/Applications/League of Legends.app/Contents/LoL/lockfile';
    } else {
      throw LcuConnectionError.unsupportedPlatform;
    }
  }
}

class LcuLockfileData {
  LcuLockfileData({required this.port, required this.password});

  final int port;
  final String password;
}

enum LcuConnectionError implements Exception {
  unsupportedPlatform,
  lockfileMissing,
  lockfileInvalid,
}
